### DISCUSSION TITLE
How to link clusters with infrastructure and apps the Kustomization way?

### DISCUSSION THREAD

**Discussion by baflo** [fluxcd/flux2#2975] [Q&A] [answered]
Hey,

I am working with fluxcd for a couple of weeks. Now I am wondering how exactly the repo is bootstrapped. After bootstrapping as described in [Get started](https://fluxcd.io/docs/get-started/), I end up with the following files:

```
└── clusters
    └── my-cluster
        └── flux-system
            ├── gotk-components.yaml
            └── gotk-sync.yaml
```

In the file `gotk-sync.yaml` there is a `Kustomization` which points to the mentioned `GitRepository` and there to the folder `./clusters/my-cluster`. However, up to this point there's not `kustomization.yaml` at that place.

In the often referenced example repository [flux2-kustomize-helm-example](https://github.com/fluxcd/flux2-kustomize-helm-example) there is neither the folder `flux-system` nor a `kustomization.yaml` in `./clusters/my-cluster`. On the other hand, in the "Get started" guide there is a `kustomization.yaml` , but it made no use of `apps` and `infrastructure` folders.

Could you give me a hint how I'd best use that `Kustomization`? Currently, I put a `kustomization.yaml` there referencing the files `apps.yaml` and `infrastructure.yaml` from the [flux2-kustomize-helm-example](https://github.com/fluxcd/flux2-kustomize-helm-example).

**Comment by kingdonb** [comment, marked as answer]
Flux is designed to build a `kustomization.yaml` for you based on `kustomize create --autodetect --recursive ...`

The behavior is described here:
https://fluxcd.io/docs/faq/#are-there-two-kustomization-types

and here is the explanation in terms of "Plain YAML" since to anyone who is not already a Kustomize user, that must all sound like some real high level wizard thing, so this should be more comprehensible to the general public who did not want to learn anything about Kustomize, we hope:
https://fluxcd.io/docs/faq/#can-i-use-repositories-with-plain-yamls

**Reply by kingdonb** [reply]
I'm afraid I missed something about your question @baflo – there was a low-hanging fruit I could have maybe answered clearer.

> In the often referenced example repository [flux2-kustomize-helm-example](https://github.com/fluxcd/flux2-kustomize-helm-example) there is neither the folder flux-system nor a kustomization.yaml in ./clusters/my-cluster. On the other hand, in the "Get started" guide there is a kustomization.yaml , but it made no use of apps and infrastructure folders.

There is no `flux-system` there because the intention is for users to fork this repo and run `flux bootstrap` on their fork. This is what the guide describes, and this would create the missing files that you're expecting to see in that repo.

When you run `flux bootstrap` that creates the GitRepository and Kustomization there, in `gotk-sync.yaml` and the Flux components in `gotk-components.yaml`. It creates a Flux Kustomization pointed at the parent directory, so you can add more folders full of manifests there and sync them together with the root `flux-system`, or you can build a tree of bases for applications in different environments like as demonstrated in `flux2-multi-tenancy`.

In the end, it's expected that advanced Flux users will have completed the Get Started guide, and both Flux2 Multi Tenancy, and Flux2 Kustomize Helm Example altogether which are listed together at the end of the Get Started guide. There's a long list of features that get introduced through those 3 examples, you would not be a proficient user of Flux and all its features without going through both advanced ones and learning enough to compose them all together and learning when the features are applicable and how they are all used.

The original intention of this structure was to make it easier for Flux users to learn all of the features without too many guides that we have to maintain them all separately. We've since gotten feedback that it's much better to have docs where task focus is the structure, and the users can start from a list of tasks and proceed directly to task they need to do. That's not the structure we've adopted so far unfortunately, but some of the links you're expecting to find between our examples appear to be links that Flux authors expected users to make for themselves. (So it seems you're on the right track here!)

**Reply by kingdonb** [reply]
I have a plan to address this with documentation/video supportive references.

I'm going to call it "Flux Bootstrap the hard way"

We can show what Bootstrap does step by step, look at every file it creates, and then follow the journey without the help of the `flux bootstrap` command. We already have some similar docs in the Azure section, because of Azure DevOps (that requires git protocol v2, which only works with libgit2, which is not yet possible to use during Bootstrap... so Azure DevOps users are already required to do bootstrap the hard way, if they want to use generic Flux, (fortunately they have the option of using the AKS Flux Extension as well))

Recognizing that I have a tendency to generate a wall of text and I might not have actually helped you here, I'm trying to think of a way that we can help more users with this common issue of not understanding how Flux works in enough detail to pick apart what is done by Kustomize and what's done by Flux, and where the line goes between them, and where users may have to add their own things etc.

Anyway, the reference from the flux2-kustomize-helm example that should have included `infrastructure` directory is here:

https://github.com/fluxcd/flux2-kustomize-helm-example#bootstrap-staging-and-production
